chrome/content/zotero/xpcom/data/item.js

```diff
@@ -4475,6 +4475,83 @@ Zotero.Item.prototype.addToCollection = function (collectionIDOrKey) {
 	this.setCollections(this._collections.concat(collectionID));
 };
 
+/**
+ * Check sync conditions for if an item can be added into a given container via drag-drop or via context menu options.
+ * @param {Zotero.Collection|Zotero.Library|Zotero.CollectionTreeRow} target - The object to which we check if an item
+ * can be added. Must be a collection, a library or a collectionTreeRow.
+ * @return {Boolean} result - true if there is no sync condition preventing an item from being added, or false otherwise.
+ */
+Zotero.Item.prototype.canAddToTarget = function (target) {
+	// If a treeRow is passed, get its ref object (which is what we care about)
+	// and record if it was a "My Publications" row
+	if (target instanceof Zotero.CollectionTreeRow) {
+		let treeRow = target;
+		target = target.ref;
+		if (treeRow.isPublications()) {
+			target.isPublications = true;
+		}
+	}
+	let targetGroup = Zotero.Libraries.get(target.libraryID);
+	if (!targetGroup.editable) return false;
+	// Target has to be a library, a collection, or "My Publications" row
+	if (!(target instanceof Zotero.Library || target instanceof Zotero.Collection || target.isPublications)) return false;
+	// Cannot move item to top-level library that it already belongs to
+	if (target instanceof Zotero.Library && target.libraryID == this.libraryID) return false;
+	// Cannot move item to a collection it already belongs to
+	if (target instanceof Zotero.Collection && target.hasItem(this)) return false;
+	// Can only move top-level items
+	if (!this.isTopLevelItem()) return false;
+	if (target.isPublications) {
+		// Can only add regular items to "My Publications"
+		if (!this.isRegularItem()) return false;
+
+		// Cannot add feed items to "My Publications"
+		if (this instanceof Zotero.FeedItem) return false;
+
+		// Cannot add item to "My Publications" that is already there
+		if (this.inPublications) return false;
+
+		// Cannot add item to "My Publications" from outside of "My Library"
+		if (this.libraryID !== Zotero.Libraries.userLibraryID) return false;
+	}
+	if (this.isAttachment() && targetGroup.libraryType == 'group') {
+		// Cannot move linked attachments to another group
+		if (this.attachmentLinkMode == Zotero.Attachments.LINK_MODE_LINKED_FILE) return false;
+
+		// Cannot move an attachment if group's files are not editable
+		if (!targetGroup.filesEditable) return false;
+	}
+	return true;
+};
```

Comment:
I assume all the `Zotero.debug()` lines were removed because they would spam the terminal when building the menu, but they're useful for debugging and when people ask why they can't copy something. Maybe add an `options` argument to these with a `debug` flag and set that to true for drags?

Reply:
Yes, that's exactly why I hadn't included them initially. But it does make sense that we want to see why something can't be copied over to a collection during drag. So I added an options `{ debug }` to `canAddToTarget` and `canAddToTargetAsync` to log why something can't be added. It's only `true` during drag-drop.